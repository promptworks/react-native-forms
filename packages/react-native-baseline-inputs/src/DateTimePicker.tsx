import * as React from "react";
import { format, parse, isValid } from "date-fns";
import { default as RNDateTimePicker } from "react-native-modal-datetime-picker";
import { DateTimePickerProps } from "./types";
import { isNil } from "./utils";
import { StaticInput } from "./StaticInput";

const parseDateTime = (value: string): Date | null => {
  const parsed = parse(value);
  return isValid(parsed) ? parsed : null;
};

const parseTime = (value: string): Date | null => {
  const today = format(new Date(), "YYYY-MM-DD");
  return parseDateTime(`${today}T${value}`);
};

const parseDateTimeLoose = (value: Date | string | null): Date | null => {
  if (typeof value === "string") {
    return parseDateTime(value) || parseTime(value);
  }

  return value;
};

interface State {
  isVisible: boolean;
}

export class DateTimePicker extends React.Component<
  DateTimePickerProps,
  State
> {
  public static defaultProps = {
    mode: "datetime",
    labelFormat: "M/D/YYYY h:mmA",
    valueFormat: "YYYY-MM-DDTHH:mm:ss.SSSZ"
  };

  public state = {
    isVisible: false
  };

  public focus() {
    this.setState({ isVisible: true });
  }

  public blur() {
    this.setState({ isVisible: false });
  }

  private handlePress = () => {
    this.focus();

    if (this.props.onFocus) {
      this.props.onFocus();
    }
  };

  private handleConfirm = (value: Date) => {
    if (this.props.onChangeDate) {
      this.props.onChangeDate(value);
    }

    if (this.props.onChange) {
      this.props.onChange(format(value, this.props.valueFormat));
    }

    this.blur();

    if (this.props.onBlur) {
      this.props.onBlur();
    }
  };

  private handleCancel = () => {
    this.blur();

    if (this.props.onBlur) {
      this.props.onBlur();
    }
  };

  public render() {
    const { isVisible } = this.state;
    const {
      value,
      style,
      labelFormat,
      inputProps,
      containerStyle,
      onFocus: _onFocus,
      onBlur: _onBlur,
      onChange: _onChange,
      onChangeDate: _onChangeDate,
      ...props
    } = this.props;

    const date = parseDateTimeLoose(value);

    return (
      <>
        <StaticInput
          onPress={this.handlePress}
          value={isNil(date) ? "" : format(date, labelFormat)}
          style={style}
          containerStyle={containerStyle}
          {...inputProps}
        />

        <RNDateTimePicker
          date={date || undefined}
          isVisible={isVisible}
          onCancel={this.handleCancel}
          onConfirm={this.handleConfirm}
          {...props}
        />
      </>
    );
  }
}
