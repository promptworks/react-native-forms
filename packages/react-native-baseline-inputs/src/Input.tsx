import * as React from "react";
import { TextInput } from "react-native";
import { InputProps } from "./types";
import { InputComponent, isNil } from "./utils";
import { styles } from "./styles";

const parse = (value: string) => {
  return /^\s*$/.test(value) ? null : value;
};

export class Input extends InputComponent<InputProps> {
  private handleChange = (value: string) => {
    if (this.props.onChange) {
      this.props.onChange(parse(value));
    }
  };

  public render() {
    const { value, onChange: _onChange, style, ...props } = this.props;

    return (
      <TextInput
        ref={this.inputRef}
        autoCapitalize="none"
        onChangeText={this.handleChange}
        style={[styles.input, style]}
        value={isNil(value) ? "" : value.toString()}
        {...props}
      />
    );
  }
}
