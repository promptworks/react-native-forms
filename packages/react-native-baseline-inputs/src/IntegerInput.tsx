import * as React from "react";
import { TextInput } from "react-native";
import { IntegerInputProps } from "./types";
import { InputComponent, isNil } from "./utils";
import { styles } from "./styles";

const clean = (input: string): string => {
  return input.replace(/[^0-9]/g, "");
};

const parse = (input: string): number | null => {
  const value = parseInt(input, 10);
  return isNaN(value) ? null : value;
};

export class IntegerInput extends InputComponent<IntegerInputProps> {
  private handleChange = (value: string) => {
    if (this.props.onChange) {
      this.props.onChange(parse(clean(value)));
    }
  };

  public render() {
    const { value, onChange: _onChange, style, ...props } = this.props;

    return (
      <TextInput
        ref={this.inputRef}
        keyboardType="numeric"
        autoCompleteType="off"
        autoCapitalize="none"
        autoCorrect={false}
        onChangeText={this.handleChange}
        style={[styles.input, style]}
        value={isNil(value) ? "" : value.toString()}
        {...props}
      />
    );
  }
}
