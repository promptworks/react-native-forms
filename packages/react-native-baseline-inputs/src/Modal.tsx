import * as React from "react";
import { ModalProps } from "./types";
import {
  Modal as RNModal,
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View
} from "react-native";

const hitSlop = {
  top: 4,
  right: 4,
  bottom: 4,
  left: 4
};

const styles = StyleSheet.create({
  backdrop: {
    flex: 1
  },
  header: {
    height: 44,
    flexDirection: "row",
    justifyContent: "flex-end",
    alignItems: "center",
    paddingHorizontal: 10,
    backgroundColor: "#EFF1F2",
    borderTopWidth: 0.5,
    borderTopColor: "#919498"
  },
  body: {
    justifyContent: "center",
    backgroundColor: "#D0D4DB"
  },
  done: {
    color: "#007AFE",
    fontWeight: "bold",
    fontSize: 15,
    paddingTop: 1,
    paddingRight: 2
  }
});

interface State {
  orientation: string;
}

export class Modal extends React.Component<ModalProps, State> {
  public state = {
    orientation: "portrait"
  };

  private handleOrientationChange = (e: NativeSyntheticEvent<any>) => {
    this.setState({ orientation: e.nativeEvent.orientation });
  };

  public render() {
    const {
      body,
      children,
      style,
      isVisible,
      headerStyle,
      backdropStyle,
      onRequestClose,
      ...modalProps
    } = this.props;

    const height = this.state.orientation === "portrait" ? 215 : 162;

    return (
      <>
        {children}

        <RNModal
          visible={isVisible}
          transparent
          animationType="slide"
          supportedOrientations={["portrait", "landscape"]}
          onRequestClose={onRequestClose}
          onOrientationChange={this.handleOrientationChange}
          {...modalProps}
        >
          <TouchableOpacity
            style={[styles.backdrop, backdropStyle]}
            onPress={onRequestClose}
          />

          <View style={[styles.header, headerStyle]}>
            <TouchableWithoutFeedback
              hitSlop={hitSlop}
              onPress={onRequestClose}
            >
              <View>
                <Text style={styles.done}>Done</Text>
              </View>
            </TouchableWithoutFeedback>
          </View>

          <View style={[styles.body, { height }, style]}>{body}</View>
        </RNModal>
      </>
    );
  }
}
