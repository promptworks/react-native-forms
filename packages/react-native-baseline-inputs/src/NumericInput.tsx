import * as React from "react";
import { TextInput } from "react-native";
import { NumericInputProps } from "./types";
import { InputComponent, isNil } from "./utils";
import { styles } from "./styles";

const clean = (input: string) => {
  return input.match(/^-?\d*\.?\d*/)![0];
};

const parse = (input: string) => {
  return ["", "-"].includes(input) ? null : input.replace(/\.$/, "");
};

export class NumericInput extends InputComponent<NumericInputProps> {
  private handleChange = (rawInput: string) => {
    const input = clean(rawInput);
    const value = parse(input);

    if (input !== rawInput) {
      this.setNativeProps({ text: input });
    }

    if (value !== this.props.value && this.props.onChange) {
      this.props.onChange(value);
    }
  };

  public render() {
    const { value, onChange: _onChange, style, ...props } = this.props;

    return (
      <TextInput
        ref={this.inputRef}
        keyboardType="numeric"
        autoCompleteType="off"
        autoCapitalize="none"
        autoCorrect={false}
        style={[styles.input, style]}
        defaultValue={isNil(value) ? "" : value}
        onChangeText={this.handleChange}
        {...props}
      />
    );
  }
}
