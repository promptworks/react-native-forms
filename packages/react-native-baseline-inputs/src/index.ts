export * from "./types";

export * from "./Input";
export * from "./NumericInput";
export * from "./IntegerInput";
export * from "./FloatInput";

export * from "./Picker";
export * from "./DatePicker";
export * from "./TimePicker";
export * from "./DateTimePicker";

export * from "./Switch";
export * from "./StaticInput";
