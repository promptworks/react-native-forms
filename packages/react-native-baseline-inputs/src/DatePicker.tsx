import * as React from "react";
import { DatePickerProps } from "./types";
import { InputComponent } from "./utils";
import { DateTimePicker } from "./DateTimePicker";

export class DatePicker extends InputComponent<DatePickerProps> {
  public static defaultProps = {
    mode: "date",
    labelFormat: "M/D/YYYY",
    valueFormat: "YYYY-MM-DD"
  };

  public render() {
    return <DateTimePicker ref={this.inputRef} {...this.props} />;
  }
}
