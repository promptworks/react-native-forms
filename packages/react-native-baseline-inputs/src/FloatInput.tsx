import * as React from "react";
import { NumericInput } from "./NumericInput";
import { FloatInputProps } from "./types";
import { InputComponent, isNil } from "./utils";

const parse = (input: string | null): number | null => {
  if (input === null) {
    return null;
  }

  const value = parseFloat(input);
  return isNaN(value) ? null : value;
};

export class FloatInput extends InputComponent<FloatInputProps> {
  private handleChange = (value: string | null) => {
    if (this.props.onChange) {
      this.props.onChange(parse(value));
    }
  };

  public render() {
    const { value, onChange: _onChange, ...props } = this.props;

    return (
      <NumericInput
        ref={this.inputRef}
        onChange={this.handleChange}
        value={isNil(value) ? "" : value.toString()}
        {...props}
      />
    );
  }
}
