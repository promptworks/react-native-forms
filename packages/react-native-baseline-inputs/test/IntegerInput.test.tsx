import * as React from "react";
import { shallow } from "enzyme";
import { IntegerInput } from "../src";

describe("<IntegerInput />", () => {
  it("renders", () => {
    const onChange = jest.fn();
    const input = shallow(<IntegerInput value={5} onChange={onChange} />);
    expect(input).toMatchSnapshot();
  });

  it("emits an integer value on change", () => {
    const onChange = jest.fn();
    const input = shallow(<IntegerInput value={null} onChange={onChange} />);
    input.find("TextInput").simulate("changeText", "aaaa1aaaaa");
    expect(onChange).toHaveBeenCalledWith(1);
  });

  it("emits a `null` value when the input cannot be parsed", () => {
    const onChange = jest.fn();
    const input = shallow(<IntegerInput value={null} onChange={onChange} />);
    input.find("TextInput").simulate("changeText", "aaaaa");
    expect(onChange).toHaveBeenCalledWith(null);
  });
});
