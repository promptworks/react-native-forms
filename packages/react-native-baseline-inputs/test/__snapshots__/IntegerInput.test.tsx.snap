// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IntegerInput /> renders 1`] = `
<TextInput
  allowFontScaling={true}
  autoCapitalize="none"
  autoCompleteType="off"
  autoCorrect={false}
  keyboardType="numeric"
  onChangeText={[Function]}
  rejectResponderTermination={true}
  style={
    Array [
      Object {
        "flex": 1,
        "fontSize": 18,
        "minHeight": 40,
      },
      undefined,
    ]
  }
  underlineColorAndroid="transparent"
  value="5"
/>
`;
